packages/tpetra/core/src/Tpetra_Details_packCrsMatrix.hpp

```diff
@@ -70,17 +72,181 @@ class Distributor;
+  typedef typename LocalMatrixType::value_type IST;
+  typedef PackCrsMatrixError value_type;
+
+  static_assert (std::is_same<LO, typename LocalMatrixType::ordinal_type>::value,
+                 "LocalMapType::local_ordinal_type and "
+                 "LocalMatrixType::ordinal_type must be the same.");
+
+  ViewType num_packets_per_lid;
+  OffsetsViewType offsets;
+  ExportsViewType exports;
+  ExportLIDsViewType export_lids;
+  LocalMatrixType local_matrix;
+  LocalMapType local_col_map;
+
+  PackCrsMatrixFunctor(ViewType num_packets_per_lid_, OffsetsViewType offsets_,
+      ExportsViewType exports_, ExportLIDsViewType export_lids_,
+      LocalMatrixType local_matrix_, LocalMapType local_col_map_):
+    num_packets_per_lid(num_packets_per_lid_), offsets(offsets_),
+    exports(exports_), export_lids(export_lids_), local_matrix(local_matrix_),
+    local_col_map(local_col_map_)
+  {}
+
+  KOKKOS_INLINE_FUNCTION void init(value_type& dst) const
+  {
+    dst.bad_index = std::numeric_limits<size_t>::max();
+    dst.bad_offset = 0;
+    dst.bad_num_bytes = 0;
+    dst.out_of_bounds_error = false;
+    dst.packing_error = false;
+  }
+
+  KOKKOS_INLINE_FUNCTION void
+  join (volatile value_type& dst, const volatile value_type& src) const
+  {
+    if (src.bad_index < dst.bad_index) {
+      dst.bad_index = src.bad_index;
+      dst.bad_offset = src.bad_offset;
+      dst.bad_num_bytes = src.bad_num_bytes;
+      dst.out_of_bounds_error = src.out_of_bounds_error;
+      dst.packing_error = src.packing_error;
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(LO i, value_type& dst) const
+  {
+    const LO export_lid = export_lids[i];
+    const size_t buf_size = exports.size();
+
+    // FIXME (mfh 24 Mar 2017) Everything here assumes UVM.  If
+    // we want to fix that, we need to write a pack kernel for
+    // the whole matrix, that runs on device.
+    size_t num_ent = static_cast<size_t>(local_matrix.graph.row_map[export_lid+1]
+                                       - local_matrix.graph.row_map[export_lid]);
+
+    // Only pack this row's data if it has a nonzero number of
+    // entries.  We can do this because receiving processes get the
+    // number of packets, and will know that zero packets means zero
+    // entries.
+
+    if (num_ent != 0) {
+
+      char* const num_ent_beg = exports.ptr_on_device() + offsets(i);
+      char* const num_ent_end = num_ent_beg + sizeof (LO);
+      char* const val_beg = num_ent_end;
+      char* const val_end = val_beg + num_ent * sizeof (IST);
+      char* const ind_beg = val_end;
+      const size_t num_bytes = num_packets_per_lid(i);
+
+      if ((offsets(i) > buf_size || offsets(i) + num_bytes > buf_size)) {
+        dst.out_of_bounds_error = true;
+      }
+      else {
+        dst.packing_error = ! packCrsMatrixRow(local_matrix, local_col_map,
+            num_ent_beg, val_beg, ind_beg, num_ent, export_lid);
+      }
+      if (dst.out_of_bounds_error || dst.packing_error) {
```

Comment:
See comments above.  I wrote the `join` and `operator` treatment of the `dst` object such that `dst` always reflected the error code and data associated with the smallest `bad_index`.